src/eq_op.rs

```diff
@@ -0,0 +1,214 @@
+use rustc::lint::*;
+use syntax::ast::*;
+use syntax::ast_util as ast_util;
+use syntax::ptr::P;
+use syntax::codemap as code;
+
+declare_lint! {
+    pub EQ_OP,
+    Warn,
+    "warn about comparing equal expressions (e.g. x == x)"
+}
+
+#[derive(Copy,Clone)]
+pub struct EqOp;
+
+impl LintPass for EqOp {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(EQ_OP)
+    }
+    
+    fn check_expr(&mut self, cx: &Context, e: &Expr) {
+        if let ExprBinary(ref op, ref left, ref right) = e.node {
+            if is_cmp_or_bit(op) && is_exp_equal(left, right) {
+                cx.span_lint(EQ_OP, e.span, &format!("equal expressions as operands to {}", ast_util::binop_to_string(op.node)));
+            }
+        }
+    }
+}
+
+fn is_exp_equal(left : &Expr, right : &Expr) -> bool {
```

Comment:
I suspect a `Folder` or a `Visitor` might be more useful for this equality check, btw.

I'm not sure if we need such an exhaustive equality check fwiw, this being a lint, but it's fine.